llamabot/bot/pydanticbot.py

```diff
@@ -0,0 +1,74 @@
+import contextvars
+
+from llamabot import SimpleBot
+from llamabot.components.messages import (
+    AIMessage, HumanMessage, SystemMessage,
+)
+from pydantic import BaseModel, ValidationError
+
+
+class PydanticBot(SimpleBot):
+    """ Pydantic Bot is given a Pydantic Model and expects the LLM to return a JSON structure that conforms to the model schema.
+        It will validate the returned json against the pydantic model, prompting the LLM to fix any of the validation errors if it does not validate, and then explicitly return an instance of that model.
+
+        This is distinct from SimpleBot's JSON-mode behaviour in the following ways:
+
+        1. JSON mode ensures a valid JSON is returned, but it does not guarantee it's in the schema you requested.
+        2. We can customize the validation rules to our needs. For example: checking for hallucinations.
+        3. Explicitly controllable number of retries.
+        4. Directly returning an instance of the Pydantic model rather than just a plain JSON string.
+    """
+
+    def __init__(
+        self,
+        system_prompt: str,
+        pydantic_model: BaseModel,
+        stream_target: str = "stdout",
+        num_attempts: int = 3,
+        **completion_kwargs,
+    ):
+        if stream_target not in ("stdout",):
+            raise ValueError(
+                f"stream_target must be one of ('stdout'), got {stream_target}."
+            )
+
+        super().__init__(system_prompt, stream_target=stream_target, json_mode=True, **completion_kwargs)
+
+        self.pydantic_model = pydantic_model
+        self.num_attempts = num_attempts
+
+    def get_system_message(self) -> SystemMessage:
+        schema = self.pydantic_model.schema_json()
+        return SystemMessage(content=self.system_prompt.content + f"\n\nYour task is to return the data in a json object that matches the following json_schema:\n```{ schema }```\nOnly return an INSTANCE of the schema, do not return the schema itself.")
```

Comment:
For this, would you be open to using the `@prompt` decorator that is present in the `llamabot.prompt_manager` submodule? It can be defined at the top of this file as a Python function:

```python
@prompt
def bot_task(schema):
    """Your task is to return the data in a json object that matches the following json_schema:

{{ schema }}

Only return an INSTANCE of the schema, do not return the schema itself."""

Then, L42 as it stands can be reformatted as:

```python
return SystemMessage(content=self.system_prompt.content + "\n\n" + bot_task(schema))
```

The separation of prompts from the code can help with readability in the long run.